### PR TITLE
Plane: Support a long failsafe that goes straight to auto

### PR DESCRIPTION
**Pull request by WickedShell**
This is very useful with an aircraft that is expected to be autonomously operating in auto from takeoff to landing. It is convenient to have a GCS connected or RC, but the loss of either isn't considered a reason to terminate the mission, and if we had interrupted for a moment it's actually better to get back into the mission.

A slightly more aggressive version of this that just went to auto regardless of flight mode has been used for several hundred flight hours on 4.2 without any issues. For actually bringing it in, I left the carveouts if we were landing, or in a Q mode.

### CODE REVIEW

**Review by khancyr — COMMENTED** [review]
we are doing it on copter, why not in plane ? 
I don't master plane code but that look in phase with other FS_ACTION handling.

**Comment by IamPete1** [comment]
This is a little odd to take the short failsafe but not the long one and then switch modes. In copter we have `FS_OPTIONS` that allows "Continue if in Auto on GCS failsafe" ect. So it would just never take the failsafe in the first place that has the advantage of not switching back into auto where the mission could be wrong. 

For example on this branch if I were to forget to upload a new mission and had a failsafe it would just fly off to the old mission and the pilot cant get it back because there in failsafe.

TLDR, I think never switching out of auto in the first place is much better than switching into auto for the long action.

**Comment by IamPete1** [comment]
I still think this is quite odd, I much prefer the "continue if in auto" concept from copter.

The premise here is that this is for vehicles that are expected to operate in auto with no RC or GCS. So the vehicle is operating in auto, the pilot then takes over in another mode (presumably because they know something the flight controller does not) then we hit failsafe and assume the pilot was wrong in there interrupt of the mission and go right back to auto.

The continue in auto would mean that a failsafe will not interrupt auto so the mission continues, but if the pilot takes over for some reason and we failsafe we then RTL (or whatever is setup) as a recovery state. 

I think this basically comes down to if we should default to trusting the pilot or default to assuming they are wrong. I can't see any safety case where this would be better. What if you need to abort your mission for some reason?

I can see that this is a better for "getting missions done", and its a option that is not enabled by default, but I would much prefer we have the  "continue if in auto" option as well so we can point the majority of people to that rather than them having to use this.

I think you can also currently put do land start at the beginning of your mission so RTL would result in the mission continuing.

**Comment by WickedShell** [comment]
> The premise here is that this is for vehicles that are expected to operate in auto with no RC or GCS

The vehicles we are using this notionally should have RC/GCS but they might be operating on the edge of the range or lose it. (Also very useful when transitioning locations.

> I think you can also currently put do land start at the beginning of your mission so RTL would result in the mission continuing.

This doesn't help if you have a mode change out, as that would send you back to the start of the mission (and also makes it harder to actually plan a real landing sequence in the mission).

> "continue if in auto" option as well so we can point the majority of people to that rather than them having to use this

I think there's a lot of value to having this option actually be useful as well, but for the use cases I've had I explicitly need auto or bust, and as said it isn't enabled by default.

File: ArduPlane/Parameters.cpp [hunk 1]
```diff
@@ -424,8 +424,8 @@ const AP_Param::Info Plane::var_info[] = {
...
-    // @Values: 0:Continue,1:ReturnToLaunch,2:Glide,3:Deploy Parachute
+    // @Description: The action to take on a long (FS_LONG_TIMEOUT seconds) failsafe event. If the aircraft was in a stabilization or manual mode when failsafe started and a long failsafe occurs then it will change to RTL mode if FS_LONG_ACTN is 0 or 1, and will change to FBWA if FS_LONG_ACTN is set to 2. If the aircraft was in an auto mode (such as AUTO or GUIDED) when the failsafe started then it will continue in the auto mode if FS_LONG_ACTN is set to 0, will change to RTL mode if FS_LONG_ACTN is set to 1 and will change to FBWA mode if FS_LONG_ACTN is set to 2. If FS_LONG_ACTN is set to 3, the parachute will be deployed (make sure the chute is configured and enabled). If FS_LONG_ACTN is set to 4 the aircraft will switch to mode auto, unless it is in the middle of a landing sequence.
```
Comment (tridge) [inline ArduPlane/Parameters.cpp:427]: should say "with the current waypoint"
also note that if already in AUTO mode then do nothing

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by peterbarker — APPROVED** [review]

**Review by tridge — APPROVED** [review]
